src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -6,8 +6,15 @@
 #####
 
 function rrule(::Type{<:Diagonal}, d::AbstractVector)
-    function Diagonal_pullback(ȳ)
-        return (NO_FIELDS, @thunk(diag(ȳ)))
+    function Diagonal_pullback(ȳ::AbstractMatrix)
+        return (NO_FIELDS, diag(ȳ))
+    end
+    function Diagonal_pullback(ȳ::Composite)
+        # TODO: Assert about the primal type in the Composite, It should be Diagonal
+        # infact it should be exactly the type of `Diagonal(d)`
+        # but right now Zygote loses primal type information so we can't use it.
+        # See https://github.com/FluxML/Zygote.jl/issues/603
```

Comment:
i am confused by this: how do we end up with a `Composite` but not a `Composite{P}` for the correct `P`?

Reply:
Because in the Zygote PR we turn all `NamedTuples` into `Composite{Any}`